### DISCUSSION TITLE
Lesson 12: FAIL. Reason: ERC20InsufficientBalance

### DISCUSSION THREAD

**Discussion by 0x-codeX** [Cyfrin/foundry-full-course-cu#1609] [Q&A] [answered]
Hi, 
I am stuck with the test below, the error says ERC20Mock has Insufficient balance and I am unclear on how to fund it so the test can go through and learn how to fix the mistakes.

I have looked inside the ERC20.sol to see where the error is from, I noticed that the balance is less than the value(amountCollateral) it is 0 from the error message. 

`function _update(address from, address to, uint256 value) internal virtual {
        if (from == address(0)) {
            // Overflow check required: The rest of the code assumes that totalSupply never overflows
            _totalSupply += value;
        } else {
            uint256 fromBalance = _balances[from];
            if (fromBalance < value) {
                revert ERC20InsufficientBalance(from, fromBalance, value);
            }`

Here is my test function.

`function testGetCollateralBalanceOfUser() public {
        vm.startPrank(user);
        ERC20Mock(weth).approve(address(engine), amountCollateral);
        engine.depositCollateral(weth, amountCollateral);
        vm.stopPrank();
        uint256 collateralBalance = engine.getCollateralBalanceOfUser(weth, user);
        assertEq(collateralBalance, amountCollateral);
    }`

error message below.

`Ran 1 test for test/unit/DSCEngineTest.t.sol:DSCEngineTest
[FAIL. Reason: ERC20InsufficientBalance(0x0000000000000000000000000000000000000001, 0, 5000000000000000000 [5e18])] testGetCollateralBalanceOfUser() (gas: 17284)
Traces:
  [9042394] DSCEngineTest::setUp()
    ├─ [4037904] → new DeployDSC@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← 20105 bytes of code
    ├─ [4777813] DeployDSC::run()
    │   ├─ [3009444] → new HelperConfig@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← ()
    │   │   ├─ [372455] → new MockV3Aggregator@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   │   └─ ← 1417 bytes of code
    │   │   ├─ [449386] → new ERC20Mock@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   │   └─ ← 2502 bytes of code
    │   │   ├─ [372455] → new MockV3Aggregator@0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3
    │   │   │   └─ ← 1417 bytes of code
    │   │   ├─ [449386] → new ERC20Mock@0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76
    │   │   │   └─ ← 2502 bytes of code
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← ()
    │   │   └─ ← 10470 bytes of code
    │   ├─ [912] HelperConfig::activeNetworkConfig() [staticcall]
    │   │   └─ ← MockV3Aggregator: [0x90193C961A926261B756D1E5bb255e67ff9498A1], MockV3Aggregator: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3], ERC20Mock: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496], ERC20Mock: [0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76], 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    │   ├─ [0] VM::startBroadcast(<pk>)
    │   │   └─ ← ()
    │   ├─ [591046] → new DecentralizedStableCoin@0x5FbDB2315678afecb367f032d93F642f64180aa3
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0x0000000000000000000000000000000000000000, newOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266)
    │   │   └─ ← 3261 bytes of code
    │   ├─ [937340] → new DSCEngine@0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
    │   │   └─ ← 5057 bytes of code
    │   ├─ [2421] DecentralizedStableCoin::transferOwnership(DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512])
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266, newOwner: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512])
    │   │   └─ ← ()
    │   ├─ [0] VM::stopBroadcast()
    │   │   └─ ← ()
    │   └─ ← DecentralizedStableCoin: [0x5FbDB2315678afecb367f032d93F642f64180aa3], DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], HelperConfig: [0x104fBc016F4bb334D775a19E8A6510109AC63E00]
    ├─ [912] HelperConfig::activeNetworkConfig() [staticcall]
    │   └─ ← MockV3Aggregator: [0x90193C961A926261B756D1E5bb255e67ff9498A1], MockV3Aggregator: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3], ERC20Mock: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496], ERC20Mock: [0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76], 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    ├─ [46789] ERC20Mock::mint(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], 10000000000000000000 [1e19])
    │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], value: 10000000000000000000 [1e19])
    │   └─ ← ()
    └─ ← ()

  [17284] DSCEngineTest::testGetCollateralBalanceOfUser()
    ├─ [0] VM::startPrank(0x0000000000000000000000000000000000000001)
    │   └─ ← ()
    ├─ [2919] ERC20Mock::transfer(DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], 5000000000000000000 [5e18])
    │   └─ ← ERC20InsufficientBalance(0x0000000000000000000000000000000000000001, 0, 5000000000000000000 [5e18])
    └─ ← ERC20InsufficientBalance(0x0000000000000000000000000000000000000001, 0, 5000000000000000000 [5e18])`

**Comment by EngrPips** [comment, marked as answer]
Hello @codexNature , it appears that the user you pranked with this line of code `vm.startPrank(user)` doesn't have enough token to deposit, make sure to mint enough token to the user before pranking them for a deposit. 

Looking forward to hear back from you.

**Reply by 0x-codeX** [reply]
Thank you @EngrPips

**Reply by EngrPips** [reply]
Awesome getting a different error message is a sign of progress, This is an assertion failure which makes more sense, when i look closely at your test you are asserting that `collateralBalance` = `amountCollateral`, and it shows in the assertion failure that `collateralBalance = 0` and `amountCollateral = 10000000000000000000` so you should probably look into your test contract and your implementation contract to see why those two values are not equal
